### K-nearest neighbors

- https://botlnec.github.io/islp/sols/chapter2/exercise7/

<img src="../images/2.7.jpg">

In [32]:
import numpy as np
import pandas as pd

In [33]:
d = {'X1': pd.Series([0,2,0,0,-1,1]),
     'X2': pd.Series([3,0,1,1,0,1]),
     'X3': pd.Series([0,0,3,2,1,1]),
     'Y': pd.Series(['Red','Red','Red','Green','Green','Red'])
    }

df = pd.DataFrame(d)
df.index = np.arange(1, len(df) + 1)
df

X1  X2  X3      Y
1   0   3   0    Red
2   2   0   0    Red
3   0   1   3    Red
4   0   1   2  Green
5  -1   0   1  Green
6   1   1   1    Red

#### (a) Euclidian distance

- Compute the Euclidean distance between each observation and the test point, X1 = X2 = X3 = 0.

In [37]:
L1 = [0, 0, 0]

#### pd.sub - element wise subtraction
df['distance'] = np.linalg.norm(df[["X1", "X2", "X3"]].sub(np.array(L1)), axis=1)
df = df.sort_values("distance")
df

X1  X2  X3      Y  distance
5  -1   0   1  Green  1.414214
6   1   1   1    Red  1.732051
2   2   0   0    Red  2.000000
4   0   1   2  Green  2.236068
1   0   3   0    Red  3.000000
3   0   1   3    Red  3.162278

#### (b) What is our prediction with K = 1? Why?

- for K=1, our prediction is Green, since that's the value of the nearest neighbor (point 5 at distance 1.41).

#### (c) What is our prediction with K = 3? Why?

- for K=3 our prediction is Red, because that's the mode of the 3 nearest neigbours: Green, Red and Red (points 5, 6 and 2, respectively).

#### (d) If the Bayes decision boundary in this problem is highly nonlinear, then would we expect the best value for K to be large or small? Why?

- A large value of K leads to a smoother decision boundary, as if the non-linearities where averaged out. This happens because KNN uses majority voting and this means less emphasis on individual points. For a large value of K, we will likely have a decision boundary which varies very little from point to point, since the result of this majority voting would have to change while for most points this will be a large majority. That is, one of its nearest neighbors changing from one class to the other would still leave the majority voting the same. By contrast, when K is very small, the decision boundary will be better able to capture local non-linearities, because it the majority of neighbors can vary significantly from point to point since the are so few neighbors. Accordingly, we would expect the best value of K to be small.